File: src/main/java/com/zin/delamain/utils/PaginationUtils.java
package com.zin.delamain.utils;

import io.javalin.http.Context;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Utility class for handling pagination across different MCP tools.
 */
public class PaginationUtils {

    private static final Logger logger = LoggerFactory.getLogger(PaginationUtils.class);

    public final int DEFAULT_PAGE_SIZE = 100;
    public final int MAX_PAGE_SIZE = 10000;
    public final int MAX_OFFSET = 1000000;

    public int getIntParam(Context ctx, String paramName, int defaultValue) {
        String param = ctx.queryParam(paramName);
        if (param == null || param.isEmpty()) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(param.trim());
        } catch (NumberFormatException e) {
            logger.warn("Invalid integer parameter '{}': '{}', using default: {}", paramName, param, defaultValue);
            return defaultValue;
        }
    }

    public <T> Map<String, Object> handlePagination(
            Context ctx,
            List<T> allItems,
            String dataType,
            String itemsKey) throws PaginationException {

        return handlePagination(ctx, allItems, dataType, itemsKey, item -> item.toString());
    }

    public <T> Map<String, Object> handlePagination(
            Context ctx,
            List<T> allItems,
            String dataType,
            String itemsKey,
            Function<T, Object> itemTransformer) throws PaginationException {

        if (allItems == null) {
            allItems = new ArrayList<>();
        }

        int totalItems = allItems.size();
        PaginationParams params = parsePaginationParams(ctx, totalItems);
        PaginationBounds bounds = calculatePaginationBounds(params, totalItems);

        List<Object> transformedItems = allItems.subList(bounds.startIndex, bounds.endIndex)
                .stream()
                .map(itemTransformer)
                .collect(Collectors.toList());

        return buildPaginationResponse(transformedItems, params, bounds, totalItems, dataType, itemsKey);
    }

    private PaginationParams parsePaginationParams(Context ctx, int totalItems) throws PaginationException {
        String offsetParam = ctx.queryParam("offset");
        String limitParam = ctx.queryParam("limit");
        String countParam = ctx.queryParam("count");

        String pageSizeParam = limitParam != null ? limitParam : countParam;

        int offset = 0;
        int requestedLimit = 0;
        boolean hasCustomLimit = pageSizeParam != null && !pageSizeParam.isEmpty();

        if (offsetParam != null && !offsetParam.isEmpty()) {
            try {
                offset = Integer.parseInt(offsetParam.trim());
                if (offset < 0) {
                    throw new PaginationException("Offset must be non-negative, got: " + offset);
                }
                if (offset > MAX_OFFSET) {
                    throw new PaginationException("Offset too large, maximum: " + MAX_OFFSET);
                }
            } catch (NumberFormatException e) {
                throw new PaginationException("Invalid offset format: '" + offsetParam + "'");
            }
        }

        if (hasCustomLimit) {
            try {
                requestedLimit = Integer.parseInt(pageSizeParam.trim());
                if (requestedLimit < 0) {
                    throw new PaginationException("Limit must be non-negative, got: " + requestedLimit);
                }
                if (requestedLimit > MAX_PAGE_SIZE) {
                    throw new PaginationException("Limit too large, maximum: " + MAX_PAGE_SIZE);
                }
            } catch (NumberFormatException e) {
                throw new PaginationException("Invalid limit format: '" + pageSizeParam + "'");
            }
        }

        int effectiveLimit;
        if (hasCustomLimit) {
            effectiveLimit = requestedLimit == 0 ? Math.max(0, totalItems - offset) : requestedLimit;
        } else {
            effectiveLimit = Math.min(DEFAULT_PAGE_SIZE, Math.max(0, totalItems - offset));
        }

        effectiveLimit = Math.max(0, Math.min(effectiveLimit, totalItems - offset));

        return new PaginationParams(offset, effectiveLimit, requestedLimit, hasCustomLimit);
    }

    private PaginationBounds calculatePaginationBounds(PaginationParams params, int totalItems) {
        if (params.offset >= totalItems) {
            return new PaginationBounds(0, 0, false, totalItems);
        }

        int startIndex = params.offset;
        int endIndex = Math.min(startIndex + params.limit, totalItems);
        boolean hasMore = endIndex < totalItems;
        int nextOffset = hasMore ? endIndex : -1;

        return new PaginationBounds(startIndex, endIndex, hasMore, nextOffset);
    }

    private Map<String, Object> buildPaginationResponse(
            List<Object> data,
            PaginationParams params,
            PaginationBounds bounds,
            int totalItems,
            String dataType,
            String itemsKey) {

        Map<String, Object> result = new HashMap<>();
        result.put("type", dataType);
        result.put(itemsKey, data);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("total", totalItems);
        pagination.put("offset", params.offset);
        pagination.put("limit", params.limit);
        pagination.put("count", data.size());
        pagination.put("has_more", bounds.hasMore);

        if (bounds.hasMore) {
            pagination.put("next_offset", bounds.nextOffset);
        }

        if (params.offset > 0) {
            int prevOffset = Math.max(0, params.offset - params.limit);
            pagination.put("prev_offset", prevOffset);
        }

        if (params.limit > 0) {
            int currentPage = (params.offset / params.limit) + 1;
            int totalPages = (int) Math.ceil((double) totalItems / params.limit);
            pagination.put("current_page", currentPage);
            pagination.put("total_pages", totalPages);
            pagination.put("page_size", params.limit);
        }

        result.put("requested_count", params.requestedLimit);
        result.put("pagination", pagination);

        return result;
    }

    private class PaginationParams {
        final int offset;
        final int limit;
        final int requestedLimit;
        final boolean hasCustomLimit;

        PaginationParams(int offset, int limit, int requestedLimit, boolean hasCustomLimit) {
            this.offset = offset;
            this.limit = limit;
            this.requestedLimit = requestedLimit;
            this.hasCustomLimit = hasCustomLimit;
        }
    }

    private class PaginationBounds {
        final int startIndex;
        final int endIndex;
        final boolean hasMore;
        final int nextOffset;

        PaginationBounds(int startIndex, int endIndex, boolean hasMore, int nextOffset) {
            this.startIndex = startIndex;
            this.endIndex = endIndex;
            this.hasMore = hasMore;
            this.nextOffset = nextOffset;
        }
    }

    public class PaginationException extends Exception {
        public PaginationException(String message) {
            super(message);
        }
    }
}

File: gateway/src/pagination_utils.py
"""
delamain Gateway - Pagination Utilities
"""

import json
import logging
from typing import Any, Callable, Dict, List, Union

logger = logging.getLogger("delamain.pagination_utils")


class PaginationUtils:
    """Utility class for handling pagination across different MCP tools."""

    DEFAULT_PAGE_SIZE = 100
    MAX_PAGE_SIZE = 10000
    MAX_OFFSET = 1000000

    @staticmethod
    def validate_pagination_params(offset: int, count: int) -> tuple[int, int]:
        if offset < 0:
            raise ValueError(f"offset must be non-negative, got: {offset}")
        if count < 0:
            raise ValueError(f"count must be non-negative, got: {count}")
        offset = min(offset, PaginationUtils.MAX_OFFSET)
        count = min(count, PaginationUtils.MAX_PAGE_SIZE)
        return offset, count

    @staticmethod
    async def get_paginated_data(
        endpoint: str,
        offset: int = 0,
        count: int = 0,
        additional_params: dict = None,
        data_extractor: Callable[[Any], List[Any]] = None,
        item_transformer: Callable[[Any], Any] = None,
        fetch_function: Callable = None,
    ) -> Union[Dict[str, Any], str]:
        try:
            offset, count = PaginationUtils.validate_pagination_params(offset, count)
        except ValueError as e:
            return {"error": f"Invalid pagination parameters: {str(e)}"}

        params = {"offset": offset}
        if count > 0:
            params["limit"] = count
        if additional_params:
            params.update(additional_params)

        try:
            if fetch_function is None:
                raise ValueError("fetch_function must be provided")

            response = await fetch_function(endpoint, params)

            try:
                status = response.get("status")
                if status in ("loading", "loading_started"):
                    return {
                        "type": response.get("type", "loading"),
                        "status": status,
                        "message": response.get("message", "Resource is loading in background"),
                        "retry_after": response.get("retry_after", 10),
                        "items": [],
                        "pagination": {"total": 0, "offset": 0, "limit": 0, "count": 0, "has_more": False},
                    }

                if "error" in response:
                    return {
                        "type": response.get("type", "error"),
                        "error": response.get("error"),
                        "suggestion": response.get("suggestion"),
                        "items": [],
                        "pagination": {"total": 0, "offset": 0, "limit": 0, "count": 0, "has_more": False},
                    }

                if data_extractor:
                    items = data_extractor(response)
                else:
                    items = (
                        response.get("classes")
                        or response.get("methods")
                        or response.get("fields")
                        or response.get("items", [])
                    )

                if item_transformer and items:
                    items = [item_transformer(item) for item in items]

                return PaginationUtils._build_standardized_response(response, items)

            except json.JSONDecodeError as e:
                logger.error(f"Failed to parse JSON response from JADX: {type(e).__name__}: {e}")
                return {"error": f"Invalid JSON response from JADX server: {type(e).__name__}"}

        except Exception as e:
            logger.error(f"Error in paginated request to {endpoint}: {type(e).__name__}: {e}")
            return {"error": f"Failed to fetch data from {endpoint}: {type(e).__name__}"}

    @staticmethod
    def _build_standardized_response(parsed_response: dict, items: List[Any]) -> dict:
        pagination_info = parsed_response.get("pagination", {})
        result = {
            "type": parsed_response.get("type", "paginated-list"),
            "items": items,
            "pagination": {
                "total": pagination_info.get("total", len(items)),
                "offset": pagination_info.get("offset", 0),
                "limit": pagination_info.get("limit", 0),
                "count": pagination_info.get("count", len(items)),
                "has_more": pagination_info.get("has_more", False),
            },
        }
        if "next_offset" in pagination_info:
            result["pagination"]["next_offset"] = pagination_info["next_offset"]
        if "prev_offset" in pagination_info:
            result["pagination"]["prev_offset"] = pagination_info["prev_offset"]
        if "current_page" in pagination_info:
            result["pagination"]["current_page"] = pagination_info["current_page"]
            result["pagination"]["total_pages"] = pagination_info.get("total_pages", 1)
            result["pagination"]["page_size"] = pagination_info.get("page_size", 0)
        return result
